---
title: "Association figures"
author: "Christian Diener"
output:
    html_notebook: default
    html_document:
        toc: true
        toc_float: true
---

# Microbiome phenotype associations

```{r}
library(mbtools)
library(ggplot2)

theme_set(theme_classic())
status_map <- c("NG", "IFG", "IGT", "IFG+ITG", "T2D", "T2D treated")

ps <- readRDS("../data/taxonomy_clean.rds")
sdata <- sample_data(ps)
sdata$diabetes_status <- factor(status_map[sdata$diabetes_status], levels=status_map)
sample_data(ps) <- sdata

annotation <- readxl::read_excel("../data/clinical_new.xlsx",
                                 sheet = "annotations_clean") %>% as.data.table()

# treatment naive samples
naive <- subset_samples(ps, as.numeric(diabetes_status) < 6 & metformin == 0)
```

Let us also get the association tests from before:

```{r}
tests <- fread("../data/tests_genus.csv")
genera <- tests[variable %in% c("diabetes_status", "num_risk_factors") 
                & padj < 0.1, names(table(genus))[table(genus) == 2]]
```

And have a look at which diabetes variables associate with which genera:

```{r, fig.width = 5, fig.height = 6}
library(pheatmap)
library(scales)

t2dsig <- tests[variable %in% c(annotation[group == "diabetes", name], "c_reactive_protein", "interleukin_6") & padj < 0.05]
mat <- dcast(t2dsig, variable ~ genus, value.var = "log2FoldChange")
rown <- mat[, variable]
mat[, variable := NULL]
mat <- as.matrix(mat)
rownames(mat) <- gsub("_", " ", rown)
mat[is.na(mat)] <- 0

maxlfc <- t2dsig[, max(abs(log2FoldChange))]
pheatmap(mat, color = colorRampPalette(c("darkblue", "white", "darkred"))(128),
         breaks = seq(-maxlfc, maxlfc, length.out = 129), 
         filename = "diabetes_sig.pdf", width = 5, height = 6)

pheatmap(mat, color = colorRampPalette(c("darkblue", "white", "darkred"))(128),
         breaks = seq(-maxlfc, maxlfc, length.out = 129))
```

## Metformin related changes

Checking for the changes reported earlier.

```{r}
data <- plot_counts(ps, "metformin", "genus", c("Escherichia/Shigella", "Intestinibacter"), 
                    normalized=T, only_data=T)

ggplot(data, aes(x=c("metformin-", "metformin+")[2 - (value == 0)], y=reads + 0.5, color=value == 0)) + 
  geom_boxplot() + geom_jitter(width=0.2, alpha=0.5) + guides(color=F) +
  facet_wrap(~ taxa) + scale_y_log10() + labs(x="", y="normalized reads")
ggsave("metformin.svg", width=6, height=4)
```

As you can see only very few samples with metformin treatment. The respective tests.

For Intestinibacter:

```{r}
wilcox.test(reads ~ value == 0, data[taxa == "Intestinibacter"])
```

For E. coli:

```{r}
wilcox.test(reads ~ value == 0, data[taxa == "Escherichia/Shigella"])
```

## Treatment-related changes

```{r, fig.width=3.5, fig.height=5.5}
t2d <- subset_samples(ps, as.numeric(diabetes_status) %in% c(1, 5, 6))

data <- plot_counts(t2d, "diabetes_status", "genus", c(genera, "Romboutsia", "Intestinibacter"), only_data=T) 
ggplot(data, aes(x=value, y=reads + 0.5, color=value)) + geom_boxplot(outlier.colour=NA) + guides(color=F) +
  geom_jitter(width=0.25, alpha=0.5, size=0.5) + scale_y_log10() + facet_wrap(~ taxa, ncol=2) + 
  labs(x="", y="normalized reads") + theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust=1))
ggsave("treated_untreated.svg", width=3.5, height=5.5)
```

Corresponding tests:

```{r}
treatment = data[, .(NG_vs_T2D = wilcox.test(reads[value == "NG"],  reads[value == "T2D"])$p.value,
                     NG_vs_treated = wilcox.test(reads[value == "NG"],  reads[value == "T2D treated"])$p.value), 
                 by=taxa]
treatment
```

## Associations with phenotype

### Diabetes group/status

```{r, fig.width=6, fig.height=6}
plot_counts(naive, "diabetes_status", "genus", genera, normalized = T) +
  geom_smooth(method="glm", aes(group=0, x=as.numeric(value))) +
  labs(x="", y="normalized reads")
ggsave("diabetes_status.svg", width=6, height=6)
```

Corresponding tests:

```{r}
tests[variable == "diabetes_status" & genus %in% genera]
```

### insulin


```{r, fig.width=6, fig.height=4}
sdata <- sample_data(naive)
ggplot(sdata, aes(x=diabetes_status, y=beta_cell_disposition_index, color=diabetes_status)) + 
  geom_hline(yintercept=2, lty="dashed") + geom_boxplot(outlier.color = NA) + 
  geom_jitter(width = 0.2, alpha = 0.5, stroke = 0) +
  labs(x="", y="beta cell disposition")
sdata$beta_cells_affected <- c("normal", "affected")[(sdata$beta_cell_disposition_index < 2) + 1]
ggsave("beta_cell_disposition.svg", width=6, height=6)

sample_data(naive) <- sdata
```

```{r, fig.width=6, fig.height=6}
data <- plot_counts(naive, "beta_cells_affected", "genus", genera, normalized = T, only_data=T)
ggplot(data[!is.na(value)], aes(x=value, y=reads, color=value)) + 
  geom_boxplot(outlier.color = NA) +
  geom_jitter(width = 0.2, alpha = 0.5, stroke = 0) + 
  facet_wrap(~ taxa) + scale_y_log10() +
  labs(x="", y="normalized reads") + guides(color=F)
ggsave("beta_cell_disposition_index.svg", width=6, height=6)
```



Corresponding tests:

```{r}
tests[variable == "beta_cell_disposition_index" & genus %in% genera]
```



```{r}
data[, .(pval = wilcox.test(reads ~ value)$p.value), by=taxa]
```


### number of risk factors:

```{r, fig.width=8, fig.height=6}
plot_counts(naive, "num_risk_factors", "genus", genera) +
  geom_smooth(method="glm", aes(group=0, x=as.numeric(value))) +
  labs(x="no. of risk factors", y="normalized reads")
ggsave("num_risk_factors.svg", width=6, height=6)
```


Corresponding tests:

```{r}
tests[variable == "num_risk_factors" & genus %in% genera]
```

We can also have a look whether the identified genera associate with the individual risk factors:

```{r, fig.width = 16, fig.height = 8}
risk_factors <- annotation[risk_factor == 1, name]
data <- lapply(risk_factors, 
               function(v) plot_counts(ps, variable = v, taxa = genera, only_data = T))
data <- rbindlist(data)

ggplot(data[!is.na(value)], aes(x=value, y=reads + 0.5, col=value)) + 
  geom_boxplot(outlier.color = NA) + 
  geom_jitter(width = 0.2, alpha = 0.5, size = 1, stroke = 0) +
  scale_y_log10() + facet_grid(taxa ~ variable, scales = "free_y") +
  guides(color = FALSE) + labs(x = "", y = "normalized reads")
ggsave("figS4.png", width = 16, height = 8, dpi = 300)
```

And the combinations that are individually significant:

```{r}
tests[variable %in% risk_factors & padj < 0.1 & genus %in% genera]
```

### Area under glucose curve

```{r, fig.width=8, fig.height=6}
data <- plot_counts(naive, "auc_glucose", "genus", genera, only_data = TRUE)
ggplot(data, aes(x = value, y = reads + 0.5)) + scale_x_log10() +
  geom_jitter(stroke = 0, alpha = 0.5) + geom_smooth(method="glm") +
  labs(x = "AUC glucose", y="normalized reads") + scale_y_log10() +
  facet_wrap(~ taxa) +
ggsave("auc_glucose.svg", width=6, height=6)
```

Corresponding tests:

```{r}
tests[variable == "auc_glucose" & genus %in% genera]
```

### Inflammation markers

Correlations with diabetes status:

```{r, fig.width=2, fig.height=3}
cor.test(sdata$c_reactive_protein, as.numeric(sdata$diabetes_status), method="spearman") %>% print()
cor.test(sdata$interleukin_6, as.numeric(sdata$diabetes_status), method="spearman") %>% print()
```

CRP in in treatment naive and treated individuals. 

```{r, fig.width=2, fig.height=3}
t2d <- subset_samples(ps, as.numeric(diabetes_status) %in% c(1, 5, 6))
t2d_sdata <- as(sample_data(t2d), "data.frame")

ggplot(t2d_sdata, aes(x=diabetes_status, y=c_reactive_protein, color=diabetes_status)) + geom_boxplot(outlier.colour=NA) + guides(color=F) +
  geom_jitter(width=0.25) + scale_y_log10() +
  labs(x="", y="CRP [mg/L]") + theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust=1))
ggsave("crp_treated_untreated.svg", width=2, height=3)

pairwise.wilcox.test(t2d_sdata$c_reactive_protein, t2d_sdata$diabetes_status, "none")
```

Interleukin 6 in treatment naive and treated individuals.

```{r, fig.width=2, fig.height=3}
ggplot(t2d_sdata, aes(x=diabetes_status, y=interleukin_6, color=diabetes_status)) + 
  geom_boxplot(outlier.colour=NA) + guides(color=F) +
  geom_jitter(width=0.25) + scale_y_log10() +
  labs(x="", y="Interleukin 6 [pg/mL]") + theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust=1))
ggsave("il6_treated_untreated.svg", width=2, height=3)

pairwise.wilcox.test(t2d_sdata$interleukin_6, t2d_sdata$diabetes_status, "none")
```

```{r, fig.width=3, fig.height=2.5}
data <- plot_counts(naive, "c_reactive_protein", "genus", "Veillonella", only_data = TRUE)
ggplot(data, aes(x = value, y = reads + 0.5)) + scale_x_log10() +
  geom_jitter(stroke = 0, alpha = 0.5) + geom_smooth(method="glm") +
  labs(x = "CRP [mg/L]", y="normalized reads") + scale_y_log10()
ggsave("c_reactive_protein.svg", width=3, height=2.5)
```

```{r, fig.width=6, fig.height=3}
data <- plot_counts(naive, "interleukin_6", "genus", c("Anaerostipes", "Blautia"), normalized = T, only_data=T)
ggplot(data, aes(x = value, y = reads + 0.5)) + scale_x_log10() +
  geom_jitter(stroke = 0, alpha = 0.5) + geom_smooth(method="glm") +
  labs(x = "Interleukin 6 [pg/mL]", y="normalized reads") + scale_y_log10() +
  facet_wrap(~ taxa)
ggsave("interleukin_6.svg", width=6, height=3)
```

Tests:

```{r}
tests[variable %in% c("c_reactive_protein", "interleukin_6") & genus %in% genera]
```

## General test stats

Significant tests from total tests:

```{r}
paste(nrow(tests[padj < 0.05]), "/", nrow(tests))
```

Genera with most significant tests:

```{r}
by_genus <- tests[, .(nsig=sum(padj < 0.05, na.rm=T)), by="genus"]
by_genus[order(-nsig)][1:10]
```

Significant tests by variable group:

```{r}
annotated_tests <- tests[annotation, on = c(variable = "name")]
by_group <- annotated_tests[, .(nsig=sum(padj < 0.05, na.rm=T),
                                ntests=length(padj),
                                nvars=uniqueN(variable)), 
                            by="group"][, "percent" := nsig/ntests]
by_group[order(-nsig)]
```

```{r}
by_genus <- annotated_tests[, .(nsig=sum(padj < 0.05, na.rm=T),
                                ntests=length(padj),
                                nvars=uniqueN(variable)), 
                            by="genus"][, "percent" := nsig/ntests]
by_genus[order(-nsig)]
```

```{r, fig.width=4, fig.height=3}
ggplot(by_group[!is.na(group)], aes(x=nsig, y=percent, color=group)) +  
  stat_smooth(aes(group = 1), method = "glm", color = "black", se = F, 
              lty = "dashed") +
  geom_point(size=2) +
  geom_text(aes(label=group), hjust=0, nudge_x=2) + xlim(0, 85) +
  labs(x="significant associations", y="positive test rate") + guides(color=F)
ggsave("sig_plot.svg", width=4, height=3)
```

```{r, fig.height=5, fig.width=8}
by_both <- annotated_tests[padj<0.05, .N, by=c("genus", "group")]
ord <- by_genus[order(-nsig), genus]
by_both[, genus := factor(genus, levels = ord)]
ggplot(by_both, aes(x=genus, y=N, fill=group)) + 
  geom_bar(stat="identity", position="stack", col="white") +
  theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5, size = 10)) +
  labs(x = "", y = "no. significant associations", fill = "")
ggsave("significant_by_genus.svg", width=8, height=5)
```
